swarm/swap/swap_test.go

```diff
@@ -0,0 +1,154 @@
+// Copyreeeeight 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package swap
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	mrand "math/rand"
+	"os"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/protocols"
+	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+	"github.com/ethereum/go-ethereum/swarm/state"
+	colorable "github.com/mattn/go-colorable"
+)
+
+var (
+	loglevel = flag.Int("loglevel", 2, "verbosity of logs")
+)
+
+func init() {
+	flag.Parse()
+
+	log.PrintOrigins(true)
+	log.Root().SetHandler(log.LvlFilterHandler(log.Lvl(*loglevel), log.StreamHandler(colorable.NewColorableStderr(), log.TerminalFormat(true))))
+}
+
+//Test that repeated bookings do correct accounting
+func TestRepeatedBookings(t *testing.T) {
+	//create a test swap account
+	swap, testDir := createTestSwap(t)
+	defer os.RemoveAll(testDir)
+
+	testPeer := newDummyPeer()
+	amount := mrand.Intn(100)
+	cnt := 1 + mrand.Intn(10)
+	for i := 0; i < cnt; i++ {
+		swap.Add(int64(amount), testPeer.Peer)
+	}
+	expectedBalance := int64(cnt * amount)
+	realBalance := swap.balances[testPeer.ID()]
+	if expectedBalance != realBalance {
+		t.Fatal(fmt.Sprintf("After %d credits of %d, expected balance to be: %d, but is: %d", cnt, amount, expectedBalance, realBalance))
+	}
+
+	testPeer2 := newDummyPeer()
+	amount = mrand.Intn(100)
+	cnt = 1 + mrand.Intn(10)
+	for i := 0; i < cnt; i++ {
+		swap.Add(0-int64(amount), testPeer2.Peer)
+	}
+	expectedBalance = int64(0 - (cnt * amount))
+	realBalance = swap.balances[testPeer2.ID()]
+	if expectedBalance != realBalance {
+		t.Fatal(fmt.Sprintf("After %d debits of %d, expected balance to be: %d, but is: %d", cnt, amount, expectedBalance, realBalance))
+	}
+
+	//mixed debits and credits
+	amount1 := mrand.Intn(100)
```

Comment:
not sure randomness here makes the test stronger :)

Reply:
👍 for not having randomness in this test.
i'd be happier if we move more in the direction of a pencil & paper kind of proofs and tests